runtime/src/main/cpp/MemorySharedRefs.cpp

```diff
@@ -33,6 +33,42 @@ RUNTIME_NORETURN inline void throwIllegalSharingException(ObjHeader* object) {
   ThrowIllegalObjectSharingException(object->type_info(), object);
 }
 
+RUNTIME_NORETURN inline void terminateWithIllegalSharingException(ObjHeader* object) {
+#if KONAN_NO_EXCEPTIONS
+  // This will terminate.
+  throwIllegalSharingException(object);
+#else
+  try {
+    throwIllegalSharingException(object);
+  } catch (...) {
+    // A trick to terminate with unhandled exception. This will print a stack trace
+    // and write to iOS crash log.
+    std::terminate();
+  }
+#endif
+}
+
+template <ErrorHandlingPolicy errorHandlingPolicy>
+bool ensureRefAccessible(ObjHeader* object, ForeignRefContext context) {
+  if (errorHandlingPolicy == ErrorHandlingPolicy::kIgnore)
+    return true;
+
+  if (isForeignRefAccessible(object, context)) {
+    return true;
+  }
+
+  switch (errorHandlingPolicy) {
```

Comment:
Consider specialization instead of switch for template parameter:
```
template <ErrorHandlingPolicy>
bool ensureRefAccessible(ObjHeader* object, ForeignRefManager* context) {
    return isForeignRefAccessible(object, context);  // default behaviour
}
// specializations:
template <>
bool ensureRefAccessible<ErrorHandlingPolicy::kIgnore>(ObjHeader* object, ForeignRefManager* context) {
    return true;
}
template <>
bool ensureRefAccessible<ErrorHandlingPolicy::kThrowException>(ObjHeader* object, ForeignRefManager* context) {
    if (isForeignRefAccessible(object, context))) return true;
    throwIllegalSharingException(object);
}
// ...
```
Looks simpler and probably more idiomatic.